vm-virtio/src/transport/pci_device.rs

```diff
@@ -797,9 +797,15 @@ impl Snapshotable for VirtioPciDevice {
         })
         .map_err(|e| MigratableError::Snapshot(e.into()))?;
 
-        let id = format!(
+        /* Use high 16 bits of device type as count of BLK/NET devices */
+        let dev_type = format!(
             "virtio-pci-{}",
-            VirtioDeviceType::from(self.device.lock().unwrap().device_type())
+            VirtioDeviceType::from(self.device.lock().unwrap().device_type() & 0xFF)
+        );
+        let id = format!(
+            "{}-{}",
+            dev_type,
+            self.device.lock().unwrap().device_type() >> 16
```

Comment:
So since this is limited to virtio devices, and since we can reasonably expect to get properly generated device IDs from the upcoming rust-vmm device manager, I'd like to make something simpler. Add an id field to `VirtioPciDevice` and atomically increase it:

```patch
diff --git a/vm-virtio/src/transport/pci_device.rs b/vm-virtio/src/transport/pci_device.rs
index 09467992..64d395db 100755
--- a/vm-virtio/src/transport/pci_device.rs
+++ b/vm-virtio/src/transport/pci_device.rs
@@ -41,6 +41,8 @@ use crate::{
     DEVICE_FEATURES_OK, DEVICE_INIT, INTERRUPT_STATUS_CONFIG_CHANGED, INTERRUPT_STATUS_USED_RING,
 };
 
+static VIRTIO_PCI_DEVICE_ID: AtomicUsize = AtomicUsize::new(0);
+
 #[allow(clippy::enum_variant_names)]
 enum PciCapabilityType {
     CommonConfig = 1,
@@ -248,6 +250,9 @@ pub struct VirtioPciDevice {
 
     // Whether to use 64-bit bar location or 32-bit
     use_64bit_bar: bool,
+
+    // Device ID
+    id: usize,
 }
 
 impl VirtioPciDevice {
@@ -339,6 +344,7 @@ impl VirtioPciDevice {
             memory: Some(memory),
             settings_bar: 0,
             use_64bit_bar,
+            id: VIRTIO_PCI_DEVICE_ID.fetch_add(1, Ordering::SeqCst),
         })
     }
 
@@ -802,11 +808,7 @@ impl Snapshotable for VirtioPciDevice {
             "virtio-pci-{}",
             VirtioDeviceType::from(self.device.lock().unwrap().device_type() & 0xFF)
         );
-        let id = format!(
-            "{}-{}",
-            dev_type,
-            self.device.lock().unwrap().device_type() >> 16
-        );
+        let id = format!("{}-{}", dev_type, self.id);
         let sec_id = format!("{}-section", id);
         let sections = vec![MigrationSection {
             id: sec_id,

```

It's simple and not intrusive at all. Until we get a properly managed set of device IDs, I think this is a better solution.